src/payment/strategies/adyenv2/adyenv2-payment-strategy.spec.ts

```diff
@@ -242,318 +116,337 @@ describe('AdyenV2PaymentStrategy', () => {
-                .mockReturnValueOnce(of(createErrorAction(PaymentActionType.SubmitPaymentFailed, identifyShopperError)))
-                .mockReturnValueOnce(submitPaymentAction);
-
-            await strategy.initialize(options);
-            await strategy.execute(getOrderRequestBody());
-
-            expect(paymentActionCreator.submitPayment).toHaveBeenCalledTimes(2);
-            expect(paymentActionCreator.submitPayment).toHaveBeenLastCalledWith({
-                methodId: 'scheme',
-                paymentData: {
-                    nonce: JSON.stringify({
-                        threeDS2Token: 'token',
-                        paymentData: 'paymentData',
+        describe('#execute', () => {
+            const identifyShopperError = getAdditionalActionError(ResultCode.IdentifyShopper);
+            const challengeShopperError = getAdditionalActionError(ResultCode.ChallengeShopper);
+            let additionalActionComponent: AdyenComponent;
+
+            beforeEach(() => {
+                let handleOnAdditionalDetails: (additionalActionState: AdyenAdditionalActionState) => {};
+
+                additionalActionComponent = {
+                    mount: jest.fn(() => {
+                        handleOnAdditionalDetails({
+                            data: {
+                                resultCode: ResultCode.ChallengeShopper,
+                                action: 'adyenAction',
+                            },
+                            isValid: true,
+                        });
+
+                        return;
                     }),
-                },
+                    unmount: jest.fn(),
+                };
+
+                jest.spyOn(adyenCheckout, 'createFromAction')
+                    .mockImplementation(jest.fn((_type, options) => {
+                        const { onAdditionalDetails } = options;
+                        handleOnAdditionalDetails = onAdditionalDetails;
+
+                        return additionalActionComponent;
+                    }));
             });
-            expect(adyenCheckout.create).toHaveBeenLastCalledWith(
-                AdyenComponentType.ThreeDS2DeviceFingerprint,
-                componentOptions
-            );
-            expect(adyenCheckout.create).toHaveBeenCalledTimes(3);
-        });
 
-        it('returns 3DS2 IdentifyShopper flow and Adyen response onError', async () => {
-            additionalActionComponent = {
-                mount: jest.fn(() => {
-                    componentOptions.onError({
-                        errorCode: 'errorCode',
-                        message: 'errorMessage',
-                    });
+            it('throws an error when payment is not present', async () => {
+                await strategy.initialize(options);
 
-                    return;
-                }),
-                unmount: jest.fn(),
-            };
+                try {
+                    await strategy.execute(getOrderRequestBodyWithoutPayment());
```

Comment:
it's better use `expect(() => ...).toThrow(PaymentArgumentInvalidError)`

The reason is, if accidentally a bug is introduced and strategy.execute stops throwing an exception, the assertion in  the "catch" block will never be executed, and the unit test won't catch the introduced bug.

```js
expect(async( ) => await strategy.execute(getOrderRequestBodyWithoutPayment()))
  .toThrowError(PaymentArgumentInvalidError));
```

another less common way is to manually call fail, but I would discourage this as it's not the pattern that we use.
```js
try {
  // code that will throw an exception
  fail();
} catch() {
 // expect..
}
```